File: typescript/src/validator/file.ts
import {
  type DataSourceOptions,
  formatDataSourceAttempts,
  resolveExistingDataSourcePath,
} from '../data-source/index.js';
import { type Diagnostic, attrValueRange, nodeRange } from '../errors/index.js';
import type {
  AttrContext,
  OpenCloseElementContext,
  SelfClosingElementContext,
} from '../generated/TDCParser.js';
import { loadFileValues } from '../generators/file.js';
import {
  buildSignalCurve,
  decimalsFromAttrs,
  parseInterp,
  parseMode,
  parsePoints,
  spreadFromAttrs,
} from '../generators/pattern.js';
import { extractAttrs } from '../processor/walk.js';
import { checkGenPattern } from './pattern.js';
import { checkSequentialDropsPercent } from './text.js';

export interface FileValidationContext {
  readonly diagnostics: Diagnostic[];
  readonly dataSources: DataSourceOptions;
}

export function checkGenFile(
  gen: OpenCloseElementContext | SelfClosingElementContext,
  ctx: FileValidationContext,
): void {
  const attrs = gen.attr();
  const attrMap = extractAttrs(attrs);
  checkSequentialDropsPercent(gen, ctx.diagnostics);
  checkQuantileRead(gen, attrMap, ctx.diagnostics);
  const srcAttr = findAttr(attrs, 'src');
  if (!srcAttr) {
    ctx.diagnostics.push({
      severity: 'error',
      source: 'validator',
      ...nodeRange(gen),
      message: '<gen type="file"> requires a "src" attribute',
      hint: 'Provide the path to a UTF-8 text file with one value per line.',
      code: 'TDC060',
    });
  }

  // Checked whether or not there is a src=, because it is not about the file:
  // a row link needs a column with or without one. Stopping at the missing src
  // would hand back one complaint, then the next on the following run, which is
  // the treadmill this whole report format exists to avoid.
  const rowAttr = findAttr(attrs, 'row');
  const columnAttr = findAttr(attrs, 'column');
  if (rowAttr && !columnAttr) {
    ctx.diagnostics.push({
      severity: 'error',
      source: 'validator',
      ...attrValueRange(rowAttr),
      message: 'row-linked file generators require a CSV "column" attribute',
      hint: 'Use column="name" or column="2" together with row="sharedKey".',
      code: 'TDC064',
    });
  }

  if (!srcAttr) return;

  const path = attrMap['src'] ?? '';
  let resolved: string;
  try {
    resolved = resolveExistingDataSourcePath(path, ctx.dataSources).path;
  } catch (err) {
    ctx.diagnostics.push({
      severity: 'error',
      source: 'validator',
      ...attrValueRange(srcAttr),
      message: `cannot read file "${path}"`,
      hint:
        err instanceof Error && 'attempts' in err
          ? formatDataSourceAttempts((err as { attempts: readonly string[] }).attempts)
          : 'Paths are relative to the current data-source base directory.',
      code: 'TDC061',
    });
    return;
  }

  if (!columnAttr) return;

  try {
    loadFileValues(resolved, {
      column: attrMap['column'],
      header: attrMap['header'],
      delimiter: attrMap['delimiter'],
    });
  } catch (err) {
    ctx.diagnostics.push({
      severity: 'error',
      source: 'validator',
      ...attrValueRange(columnAttr),
      message: err instanceof Error ? err.message : String(err),
      hint: 'For CSV files, use a header name like column="email" or a 1-based index like column="2".',
      code: 'TDC062',
    });
  }
}

/** The three ways a `<gen type="pattern">` can be given a shape to read. */
const DRAWING_SOURCES = ['points', 'src', 'upper'] as const;

/**
 * A drawing's `src=` — the same existence check, without the file generator's
 * own rules about columns and linked rows.
 *
 * Checked before the run rather than during it: a missing picture discovered on
 * row one of a million-row job has already cost whatever the job cost. `src=` is
 * one of three ways to hand the generator a shape, so its absence is only a
 * mistake when the other two are absent too — which is TDC244, the drawing
 * equivalent of a `type="regex"` with no pattern.
 */
export function checkGenDrawing(
  gen: OpenCloseElementContext | SelfClosingElementContext,
  ctx: FileValidationContext,
): void {
  const attrs = gen.attr();
  const attrMap = extractAttrs(attrs);
  checkGenPattern(gen, ctx.diagnostics);
  if (DRAWING_SOURCES.every((name) => (attrMap[name] ?? '').trim() === '')) {
    ctx.diagnostics.push({
      severity: 'error',
      source: 'validator',
      ...nodeRange(gen),
      message: '<gen type="pattern"> has nothing to draw from',
      hint: 'Give it a shape: points="0,0 1,5 2,3", src="curve.svg" (or a PNG), or upper="…" with an optional lower="…" for a band.',
      code: 'TDC244',
    });
    return;
  }

  checkDrawingValues(attrs, attrMap, ctx.diagnostics);

  const srcAttr = findAttr(attrs, 'src');
  const path = attrMap['src'] ?? '';
  if (!srcAttr || path.trim() === '') return;

  try {
    resolveExistingDataSourcePath(path, ctx.dataSources);
  } catch (err) {
    ctx.diagnostics.push({
      severity: 'error',
      source: 'validator',
      ...attrValueRange(srcAttr),
      message: `cannot read file "${path}"`,
      hint:
        err instanceof Error && 'attempts' in err
          ? formatDataSourceAttempts((err as { attempts: readonly string[] }).attempts)
          : 'Paths are relative to the current data-source base directory.',
      code: 'TDC061',
    });
  }
}

function findAttr(attrs: readonly AttrContext[], name: string): AttrContext | undefined {
  for (const attr of attrs) {
    if (attr._attrName?.text === name) return attr;
  }
  return undefined;
}

/**
 * `mode=`, `interp=`, `spread=` and `decimals=` — the four drawing attributes whose
 * value is a fixed word or a number.
 *
 * They were read only by the generator, so `check` called `mode="banana"` valid and
 * the run then refused it with a bare sentence and no code: `check` answers "would
 * this run?" everywhere else, and here it answered wrongly for a config a reader
 * would fix in a second.
 *
 * The validator calls the GENERATOR's own parsers rather than repeating their rules.
 * A second copy of "linear, smooth or step" is a second thing to keep in step, and the
 * failure that produces — `check` accepting what the run refuses — is exactly the one
 * being closed here.
 */
function checkDrawingValues(
  attrs: readonly AttrContext[],
  attrMap: Record<string, string>,
  diagnostics: Diagnostic[],
): void {
  const checks: readonly [string, () => unknown][] = [
    // The three that carry a DRAWING, checked by the same reader the run uses. A `;`
    // that reads as one curve instead of two, and a set of points with no width, both
    // used to pass `check` and die in the run.
    ['points', () => buildSignalCurve(parsePoints(attrMap['points'] ?? ''), undefined, 0)],
    ['upper', () => buildSignalCurve(parsePoints(attrMap['upper'] ?? ''), undefined, 0)],
    ['lower', () => buildSignalCurve(parsePoints(attrMap['lower'] ?? ''), undefined, 0)],
    ['mode', () => parseMode(attrMap['mode'])],
    ['interp', () => parseInterp(attrMap['interp'])],
    ['spread', () => spreadFromAttrs(attrMap)],
    ['decimals', () => decimalsFromAttrs(attrMap)],
  ];
  for (const [name, run] of checks) {
    const attr = findAttr(attrs, name);
    if (!attr) continue;
    try {
      run();
    } catch (err) {
      diagnostics.push({
        severity: 'error',
        source: 'validator',
        ...attrValueRange(attr),
        message: err instanceof Error ? err.message : String(err),
        hint: 'Every drawing attribute is checked before the run, so `check` and the run agree.',
        code: 'TDC285',
      });
    }
  }
}

/**
 * `read="quantile"` — the file as a sorted sample rather than a bag of values.
 *
 * Everything refused here is a config that asks for TWO readings of one file at
 * once. `weight=` says "the shares live in a column beside the values";
 * `read="quantile"` says "the values ARE the distribution, and how often each
 * appears is its share". `row=` links several columns to one LINE of the file,
 * and a quantile answer is not a line — it is a point between two of them.
 * `order="sequential"` walks the list in order, which is the one thing a
 * distribution has no notion of.
 *
 * Each of these would otherwise have been resolved by whichever branch the
 * engine reached first, silently — the defect this project keeps closing.
 */
function checkQuantileRead(
  gen: OpenCloseElementContext | SelfClosingElementContext,
  attrMap: Record<string, string>,
  diagnostics: Diagnostic[],
): void {
  const attrs = gen.attr();
  const readAttr = findAttr(attrs, 'read');
  const read = (attrMap['read'] ?? '').trim();
  const sampleAttr = findAttr(attrs, 'sample');
  const sample = (attrMap['sample'] ?? '').trim();

  if (readAttr && read !== 'quantile') {
    diagnostics.push({
      severity: 'error',
      source: 'validator',
      ...attrValueRange(readAttr),
      message: `read="${read}" is not a way of reading a file — the only one is "quantile"`,
      hint: 'Leave read= off to pick one of the file\'s values at random, or write read="quantile" to read the file as a sorted sample and land anywhere on it.',
      code: 'TDC297',
    });
    return;
  }

  if (sampleAttr && sample !== 'exact') {
    diagnostics.push({
      severity: 'error',
      source: 'validator',
      ...attrValueRange(sampleAttr),
      message: `sample="${sample}" is not a sampling mode — the only one is "exact"`,
      hint: 'Leave sample= off to draw from the distribution row by row, or write sample="exact" to sweep it evenly so the run reproduces the sample with no sampling noise.',
      code: 'TDC297',
    });
  }

  if (sampleAttr && read !== 'quantile') {
    diagnostics.push({
      severity: 'error',
      source: 'validator',
      ...attrValueRange(sampleAttr),
      message: 'sample= only means something beside read="quantile"',
      hint: 'It chooses between drawing from the distribution and sweeping it evenly, and a file read as a plain list of values has no distribution to sweep.',
      code: 'TDC297',
    });
  }

  if (read !== 'quantile') return;

  for (const [name, why] of [
    [
      'weight',
      'weight= puts the shares in a COLUMN beside the values, and read="quantile" says the values are the distribution themselves — how often one appears in the file IS its share',
    ],
    [
      'row',
      'row= links several columns to one LINE of the file, and a quantile answer is not a line: it is a point between two of them',
    ],
  ] as const) {
    const attr = findAttr(attrs, name);
    if (!attr) continue;
    diagnostics.push({
      severity: 'error',
      source: 'validator',
      ...attrValueRange(attr),
      message: `${name}= cannot be combined with read="quantile": ${why}`,
      hint: `Keep one of the two readings. ${name === 'weight' ? 'A countable value — a city, a status, a number of orders — wants weight= and its exact quota; a measured one wants the quantile read, which also fills in the values between the observations.' : 'To keep a record together, read the file as lines with row= and leave read= off.'}`,
      code: 'TDC297',
    });
  }

  if ((attrMap['order'] ?? '').trim() === 'sequential') {
    const attr = findAttr(attrs, 'order');
    diagnostics.push({
      severity: 'error',
      source: 'validator',
      ...(attr ? attrValueRange(attr) : nodeRange(gen)),
      message: 'order="sequential" cannot be combined with read="quantile"',
      hint: "Walking a list in order and sampling a distribution are different jobs: one hands out the file's lines one after another, the other says where on the sorted sample a row lands.",
      code: 'TDC297',
    });
  }
}

File: python/src/tdcv2/generators/file.py
"""``<gen type="file" src="./products.txt"/>`` — values from the user's own file.

Two shapes. A plain list is one value per line, blanks skipped. With ``column=`` the file is read
as CSV and one column is taken from it — by header name, or by 1-based position when the column is
written as a number.

This is how a run gets the real thing: the actual product catalogue, the actual list of branch
codes. Generated data is only as convincing as the vocabulary it draws from, and no bundled pack
knows one particular company's part numbers.
"""

from __future__ import annotations

import os
import re
from dataclasses import dataclass
from pathlib import Path
from urllib.parse import urlparse
from urllib.request import url2pathname

from ..prng import rand
from ..prng.prng import Sfc32

_NUMBERED = re.compile(r"^[1-9][0-9]*$")
_LINE_BREAK = re.compile(r"\r?\n")

# Excel writes one of these ahead of the first header cell. JavaScript's `trim` counts it as
# whitespace and Python's `strip` does not, so it comes off by name.
_BOM = "﻿"


@dataclass(frozen=True, slots=True)
class Weighted:
    """Values and their shares, when ``weight="countColumn"`` names a second column.

    Without it a list is drawn uniformly, so Smith and Zabrowski turn up equally often. Real
    distributions are never flat — the commonest surnames cover a large part of a population and
    the tail is vanishingly rare — and flattening that is the first thing anyone looking at the
    data notices.

    The shares are honoured exactly, through the same apportionment ``percent=`` uses: weights of
    20000 and 10000 over 30000 rows give precisely 20000 and 10000, not "about twice as many". A
    weight is a raw count, not a percentage, because census and registry files publish counts and
    normalising them by hand is a pointless invitation to arithmetic errors.
    """

    values: list[str]
    percents: list[float]


@dataclass(frozen=True, slots=True)
class RowSource:
    """A CSV read as whole rows, for ``row="key"``.

    Several sequences naming the same key read different columns of the SAME row, so a generated
    city and its postcode come from one real record rather than from two unrelated ones. Without
    it, drawing a city and a postcode independently produces pairs that no validator and no human
    would accept.
    """

    rows: list[list[str]]
    header: list[str]
    column_index: int
    source_key: str


def generate(
    attrs: dict[str, str],
    count: int,
    base_dir: Path | None,
    prng: Sfc32,
    roots: list[Path] | None = None,
) -> list[str]:
    values = load(attrs, base_dir, roots)
    return [rand.pick(prng, values) for _ in range(count)]


def load(
    attrs: dict[str, str], base_dir: Path | None, roots: list[Path] | None = None
) -> list[str]:
    """The file's values in file order — what ``order="sequential"`` reads."""
    src = attrs.get("src")
    if src is None or not src.strip():
        raise ValueError('file generator: "src" is required')
    path = resolve(src.strip(), base_dir, roots)
    content = _read(path)

    column = _trim_to_none(attrs.get("column"))
    if "column" in attrs and column is None:
        raise ValueError("file generator: column must not be empty")

    values = _list_values(content) if column is None else _csv_column(content, column, attrs, path)
    if not values:
        raise ValueError(f'file generator: list at "{path}" is empty')
    return values


def load_weighted(
    attrs: dict[str, str], base_dir: Path | None, roots: list[Path] | None = None
) -> Weighted | None:
    """``None`` when the generator is not weighted, which is the ordinary case."""
    weight_column = _trim_to_none(attrs.get("weight"))
    if weight_column is None:
        return None
    column = _trim_to_none(attrs.get("column"))
    if column is None:
        raise ValueError(f'file generator: weight="{weight_column}" needs a "column" to weight')

    path = resolve(attrs["src"].strip(), base_dir, roots)
    rows = _non_blank_rows(_read(path), parse_delimiter(attrs.get("delimiter")))
    if not rows:
        raise ValueError(f'file generator: CSV file at "{path}" is empty')

    header = rows[0]
    value_index = column_index(header, column)
    weight_index = column_index(header, weight_column)
    if weight_index == value_index:
        raise ValueError(
            f'file generator: weight column "{weight_column}" is the same column as the values'
        )

    values: list[str] = []
    counts: list[float] = []
    for position, row in enumerate(rows[1:]):
        value = _cell(row, value_index)
        # The same refusal the plain path makes, for the same reason.
        if not value:
            raise ValueError(
                f'file generator: column "{column}" is empty on value row {position + 1} '
                f'— a blank cell would drop that row from the values and quietly change the proportions. Fill it in, remove the row, or point column= at a column that is complete.'
            )
        weight = _weight_of(row, weight_index, weight_column, value)
        # A zero weight means never drawn, so carrying it costs memory and buys nothing.
        if weight == 0:
            continue
        values.append(value)
        counts.append(weight)

    if not values:
        raise ValueError(
            f'file generator: no values with a positive weight in column "{weight_column}"'
        )
    return Weighted(values, _as_percents(counts))


def load_rows(
    attrs: dict[str, str], base_dir: Path | None, roots: list[Path] | None = None
) -> RowSource:
    column = _trim_to_none(attrs.get("column"))
    if column is None:
        raise ValueError('sequence: row-linked file generator requires a CSV "column" attribute')
    src = (attrs.get("src") or "").strip()
    path = resolve(src, base_dir, roots)
    delimiter = parse_delimiter(attrs.get("delimiter"))

    everything = _non_blank_rows(_read(path), delimiter)
    if not everything:
        raise ValueError(f'file generator: CSV file at "{src}" is empty')

    index = column_index(everything[0], column)
    skip_header = _parse_header_flag(attrs.get("header")) or not _NUMBERED.match(column)
    rows = everything[1:] if skip_header else everything

    if not rows:
        raise ValueError(f'file generator: CSV file at "{src}" has no data rows')
    if not any(_cell(row, index) for row in rows):
        raise ValueError(f'file generator: CSV column "{column}" at "{src}" has no values')

    # The header is kept: `rows` may have had it stripped, so a second column named later — a
    # weight column — has to be resolved against the original.
    #
    # Two sequences on one key must be reading one file; source_key identifies which.
    return RowSource(rows, everything[0], index, f"{path}|{delimiter}|{str(skip_header).lower()}")


def cell_at(source: RowSource, row_index: int) -> str:
    """One row's cell in the linked column, trimmed and never absent."""
    return _cell(source.rows[row_index], source.column_index)


def weighted_rows(attrs: dict[str, str], source: RowSource) -> Weighted | None:
    """Row indexes drawn to the exact quota of a weight column.

    The row-linked counterpart of :func:`load_weighted`, so every field on the link follows the
    same weighted rows.
    """
    weight_column = _trim_to_none(attrs.get("weight"))
    if weight_column is None:
        return None
    weight_index = column_index(source.header, weight_column)
    if weight_index == source.column_index:
        raise ValueError(
            f'file generator: weight column "{weight_column}" is the same column as the values'
        )

    indexes: list[str] = []
    counts: list[float] = []
    for i, row in enumerate(source.rows):
        value = cell_at(source, i)
        if not value:
            continue
        weight = _weight_of(row, weight_index, weight_column, value)
        if weight == 0:
            continue
        indexes.append(str(i))
        counts.append(weight)

    if not indexes:
        raise ValueError(
            f'file generator: weight column "{weight_column}" has no rows with a positive weight'
        )
    return Weighted(indexes, _as_percents(counts))


#: The prefix that says "look in the configured data folders, not next to the config".
DATA_ALIAS = "@data/"


def resolve(src: str, base_dir: Path | None, roots: list[Path] | None = None) -> Path:
    """Where a ``src=`` points, in the order the reference implementation looks.

    A plain relative path means the file next to the CONFIG, not next to whatever directory the
    program happened to be started from — otherwise the same config would work from one shell and
    fail from another. When it is not there, the configured data folders are tried, so a config can
    be moved without rewriting every source.

    ``@data/x.txt`` skips the config's folder entirely and names the data folders outright. That is
    what makes a config portable between machines whose data lives in different places: the path in
    the config stays the same and only ``--data-path`` (or ``dataPaths`` in the project config)
    differs. With no data folder configured at all the alias cannot mean anything, and saying so is
    better than reporting a missing file.
    """
    text = src.strip()
    candidates = list(roots or [])

    if text.startswith("file://"):
        return Path(url2pathname(urlparse(text).path))

    if text.startswith(DATA_ALIAS):
        alias = text[len(DATA_ALIAS) :].strip()
        if not alias:
            raise ValueError("file generator: @data source path must not be empty")
        if not candidates:
            raise ValueError(
                'file generator: "@data/..." needs at least one data folder — '
                "pass --data-path, or name one in tdcv2.config.json"
            )
        return _first_readable([root / alias for root in candidates])

    path = Path(text)
    if path.is_absolute():
        return path

    beside = Path(os.path.normpath(base_dir / path)) if base_dir is not None else path
    if beside.is_file() or not candidates:
        return beside
    return _first_readable([beside, *(root / path for root in candidates)])


def _first_readable(attempts: list[Path]) -> Path:
    """The first candidate that exists, or the first one tried so the error names something real."""
    for candidate in attempts:
        if candidate.is_file():
            return candidate
    return attempts[0]


def parse_rows(content: str, delimiter: str = ",") -> list[list[str]]:
    """RFC 4180: quoted fields, doubled quotes inside them, and either line ending."""
    rows: list[list[str]] = []
    row: list[str] = []
    field: list[str] = []
    in_quotes = False
    quoted_field = False

    i = 0
    length = len(content)
    while i < length:
        ch = content[i]
        if in_quotes:
            if ch == '"':
                if i + 1 < length and content[i + 1] == '"':
                    field.append('"')
                    i += 1
                else:
                    in_quotes = False
            else:
                field.append(ch)
            i += 1
            continue

        if ch == '"' and not field and not quoted_field:
            in_quotes = True
            quoted_field = True
        elif ch == delimiter:
            row.append("".join(field))
            field = []
            quoted_field = False
        elif ch in ("\n", "\r"):
            if ch == "\r" and i + 1 < length and content[i + 1] == "\n":
                i += 1
            row.append("".join(field))
            field = []
            quoted_field = False
            rows.append(row)
            row = []
        else:
            field.append(ch)
        i += 1

    if in_quotes:
        raise ValueError("file generator: unterminated quoted CSV field")
    if field or row or not content.endswith("\n"):
        row.append("".join(field))
        rows.append(row)
    return rows


def parse_delimiter(value: str | None) -> str:
    if value is None:
        return ","
    # A single character is taken as written, tab included, so that resolving twice is harmless:
    # trimming a real tab would leave nothing and fall back to a comma.
    if len(value) == 1:
        return value
    normalized = value.strip()
    if not normalized:
        return ","
    aliases = {"comma": ",", "semicolon": ";", "tab": "\t", "\\t": "\t", "pipe": "|"}
    resolved = aliases.get(normalized.lower(), normalized)
    if len(resolved) != 1:
        raise ValueError("file generator: delimiter must be one character")
    return resolved


def column_index(header_row: list[str], column: str) -> int:
    if _NUMBERED.match(column):
        return int(column) - 1
    for i, cell in enumerate(header_row):
        if cell.replace(_BOM, "").strip() == column:
            return i
    raise ValueError(f'file generator: CSV column "{column}" was not found in the header row')


def _read(path: Path) -> str:
    try:
        return path.read_text(encoding="utf-8")
    except OSError as e:
        raise ValueError(f'file generator: cannot read "{path}"') from e


def _list_values(content: str) -> list[str]:
    return [line for line in (raw.strip() for raw in _LINE_BREAK.split(content)) if line]


def _csv_column(content: str, column: str, attrs: dict[str, str], path: Path) -> list[str]:
    rows = _non_blank_rows(content, parse_delimiter(attrs.get("delimiter")))
    if not rows:
        raise ValueError(f'file generator: CSV file at "{path}" is empty')

    index = column_index(rows[0], column)
    # A named column implies a header row; a numbered one only skips it when told to, because a
    # file of pure data has no header to skip.
    skip_header = _parse_header_flag(attrs.get("header")) or not _NUMBERED.match(column)

    # A blank cell is REFUSED, not skipped. Dropping it takes the row out of the pool, so the
    # file's own proportions stop being the run's: measured on a three-person CSV with one
    # empty email, 60 rows produced 28 and 32 of the other two and no sign of the third. The
    # weighted path refuses the same shape one column over, for the same stated reason.
    values = [_cell(row, index) for row in rows[1 if skip_header else 0 :]]
    for position, cell in enumerate(values):
        if not cell:
            raise ValueError(
                f'file generator: column "{column}" is empty on value row {position + 1} of '
                f'"{path}" — a blank cell would drop that row from the values and quietly change the proportions. Fill it in, remove the row, or point column= at a column that is complete.'
            )
    if not values:
        raise ValueError(f'file generator: CSV column "{column}" at "{path}" has no values')
    return values


def _non_blank_rows(content: str, delimiter: str) -> list[list[str]]:
    return [row for row in parse_rows(content, delimiter) if any(cell.strip() for cell in row)]


def _cell(row: list[str], index: int) -> str:
    return row[index].strip() if index < len(row) else ""


def _weight_of(row: list[str], index: int, weight_column: str, value: str) -> float:
    raw = _cell(row, index)
    # A blank cell must not slide through as a weight of zero, which would delete the value from
    # the run. A product vanishing from a catalogue because one cell of an export was empty is
    # discovered far too late — and missing data and a deliberate zero are different statements,
    # only one of which is actionable.
    if not raw:
        raise ValueError(
            f'file generator: weight column "{weight_column}" is empty for value "{value}" '
            "— write 0 to exclude it, or fill in the count"
        )
    try:
        # Python alone reads "1_000" as a thousand. A weight column is data from someone else's
        # export, and the three implementations have to reject the same cells.
        weight = float("nan") if "_" in raw else float(raw)
    except ValueError:
        weight = float("nan")
    if weight != weight or weight in (float("inf"), float("-inf")) or weight < 0:
        raise ValueError(
            f'file generator: weight "{raw}" for value "{value}" is not a non-negative number'
        )
    return weight


def _as_percents(counts: list[float]) -> list[float]:
    total = sum(counts)
    return [count / total * 100 for count in counts]


def _parse_header_flag(value: str | None) -> bool:
    if value is None:
        return False
    normalized = value.strip().lower()
    if normalized in ("true", "1"):
        return True
    if normalized in ("false", "0"):
        return False
    raise ValueError("file generator: header must be true or false")


def _trim_to_none(value: str | None) -> str | None:
    if value is None:
        return None
    trimmed = value.strip()
    return trimmed or None
